Force collections to be static

Shrews opened the pull request:
Templating of collection names does not work at all. Force them to
be static so that a warning is generated for the user.

##### SUMMARY

Attempting to template a collection name within a playbook results in any of the collection roles/modules/etc not being found and generates an error. This will force the collection to be a static attribute, disallowing templating, and generating a warning message for the user.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

collections

##### ADDITIONAL INFORMATION

Closes #68704

Shrews committed:
Force collections to be static

Templating of collection names does not work at all. Force them to
be static so that a warning is generated for the user.

sivel commented on lib/ansible/playbook/collectionsearch.py:
```diff
@@ -32,9 +36,18 @@ def _ensure_default_collection(collection_list=None):
+        # because if the user attempts to template a collection name, it will
+        # error before it ever gets to the post_validate() warning.
+        for collection_name in ds:
+            if is_template(collection_name, Environment()):
```
Instantiate `Environment` 1 time and re-use it.

To address other potential questions here, we don't allow for things like `variable_start` to be overriden in the playbook "DSL", as such, we don't need to account for that here.  Although we could in the future, but I'm not concerned.

Also, `Environment.lex` doesn't care about undefined vars, or non-existent filters, etc...So we don't need that context anyway.

sivel replied:
I think this code may actually need to be moved after `_ensure_default_collection`.  `_ensure_default_collection` will make sure the value is formatted correctly.  In the case where someone just does `collections: whatever` it will be turned into a list as well.  So maybe it needs moved all the way to the bottom.

Shrews replied:
Doh! on the Environment usage. All requested changes applied.

sivel commented:
For a unit test, you can probably just do something like:

```python
from ansible.playbook.collectionsearch import CollectionSearch
cs = CollectionSearch()
assert cs._load_collections(None, ['foo.{{bar}}']) ==  ['foo.{{bar}}']
```

And then also monkeypatch/mock `Display`, and assert that it was called as expected.

Shrews committed:
Add collectionsearch unit test and fix for reviews

New unit test validates the new _load_collections() code and moves
the new check to the end of the method.

webknjaz commented on test/units/playbook/test_collectionsearch.py:
```diff
@@ -0,0 +1,42 @@
+from units.compat.mock import patch
+
+
+class TestCollectionSearch(unittest.TestCase):
```
Inheriting from `unittest.TestCase` is harmful b/c our testrunner is `pytest` and this conflicts with a part of its features.

webknjaz commented on test/units/playbook/test_collectionsearch.py:
```diff
@@ -0,0 +1,42 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from units.compat import unittest
+from ansible.playbook.collectionsearch import CollectionSearch
+from units.compat.mock import patch
+
+
+class TestCollectionSearch(unittest.TestCase):
+
+    @patch('ansible.utils.display.Display.display')
+    def test_collection_static_warning(self, mock_display):
+        '''
+        Test that collection name is not templated and a warning message is
+        generated for the referenced name.
+        '''
+        cs = CollectionSearch()
+        self.assertIn(
+            'foo.{{bar}}',
+            cs._load_collections(None, ['foo.{{bar}}'])
+        )
+        self.assertEqual(mock_display.call_count, 1)
+        self.assertIn(
+            '[WARNING]: "collections" is not templatable, but we found: foo.{{bar}}',
+            mock_display.call_args[0][0])
```
I know you've been advised to use `pytest-mock` and [`monkeypatch`](https://docs.pytest.org/en/latest/monkeypatch.html) fixtures but turns out you don't actually need these.

Pytest already captures I/O file descriptors. So what you need is to use what's recorded. This way it's closer to e2e and you don't do unnecessary patching of the logic.

There's [a fixture for capturing warnings `recwarn`](https://docs.pytest.org/en/latest/warnings.html) but it's not what we need here.

Instead, we need [`capsys`](https://docs.pytest.org/en/latest/capture.html#accessing-captured-output-from-a-test-function) (or one of its variations).

And in the end, this test would look like:
```suggestion
from ansible.playbook.collectionsearch import CollectionSearch

import pytest


def test_collection_static_warning(capsys):
    """Test that collection name is not templated.

    Also, make sure that users see the warning message for the referenced name.
    """
    cs = CollectionSearch()
    assert 'foo.{{bar}}' in cs._load_collections(None, ['foo.{{bar}}'])

    std_out, std_err = capsys.readouterr()
    assert '[WARNING]: "collections" is not templatable, but we found: foo.{{bar}}\n' == std_err
    assert '' == std_out
```
_(DISCLAIMER: this is coded in browser, you can accept it by clicking the `Commit suggestion` UI button under this patch ^ and see if I haven't made any typos)_

webknjaz commented on test/units/playbook/test_collectionsearch.py:
```diff
@@ -0,0 +1,42 @@
+    @patch('ansible.utils.display.Display.display')
+    def test_collection_static_warning(self, mock_display):
+        '''
+        Test that collection name is not templated and a warning message is
```
It's best to follow PEP257 for docstrings.

Shrews committed:
Change unit test to pytest

webknjaz reviewed (dismissed):
No need to patch things, see https://github.com/ansible/ansible/pull/68723#discussion_r404821814

Shrews committed:
Adjust unit test to use capsys instead of monkeypatch

webknjaz commented on test/units/playbook/test_collectionsearch.py:
```diff
@@ -0,0 +1,39 @@
+    assert mock_display.call_count == 1
+    actual_warn = ' '.join(mock_display.mock_calls[0][1][0].split('\n'))
+    expected_warn = '"collections" is not templatable, but we found: %s' \
+        % to_text(collection_name)
```
It's probably not the best idea to render a Unicode object into the literal that can be unicode or bytes depending on the interpreter version... I'd probably use `no_native()` here.

webknjaz commented on test/units/playbook/test_collectionsearch.py:
```diff
@@ -0,0 +1,39 @@
+__metaclass__ = type
+
+from ansible.playbook.collectionsearch import CollectionSearch
+from ansible.module_utils._text import to_bytes, to_native, to_text
```
This has 2 unused imports

webknjaz commented on test/units/playbook/test_collectionsearch.py:
```diff
@@ -0,0 +1,38 @@
+    assert collection_name in cs._load_collections(None, [collection_name])
+
+    std_out, std_err = capsys.readouterr()
+    assert '[WARNING]: "collections" is not templatable, but we found: %s' % collection_name in std_err
```
I'd use equality which would be more precise and replace that check that you had earlier with calls count == 1.

Shrews replied:
I was using IN here because it seems the capsys output contains color-coding chars and newlines so IN was easier to use for confirming the main data here (warning + collection name) is present.

webknjaz replied:
Right, you could probably monkeypatch `ANSIBLE_COLOR`: https://github.com/ansible/ansible/blob/598ead965dd0ed1db9a554e8a02c8fcc693ffdd8/lib/ansible/utils/color.py#L91.

But I looked closer at the display function and it looks like it uses an actual logger from `logging` stdlib module (https://github.com/ansible/ansible/blob/598ead965dd0ed1db9a554e8a02c8fcc693ffdd8/lib/ansible/utils/display.py#L216). So it's even easier than I thought initially.

You should be able to use the [`caplog` fixture](https://docs.pytest.org/en/latest/logging.html#caplog-fixture). Let me see if I can quickly come up with a patch.

webknjaz commented on test/units/playbook/test_collectionsearch.py:
```diff
@@ -0,0 +1,38 @@
+
+    collection_name = 'foo.{{bar}}'
+    cs = CollectionSearch()
+    assert collection_name in cs._load_collections(None, [collection_name])
```
What does that method return? Can we have an equality check here?
Also, is there a public method that could be called instead of the private one?

Shrews replied:
It returns a list of collection names and is adding 'ansible.legacy' to that list. I'm not aware of a public method to call instead.

webknjaz replied:
If the order is deterministic, you could have a more precise check here.

Shrews replied:
Well, in this test case, it should always be the first element based on the current code. But in this case, we don't actually care about positioning, just that it appears anywhere in the list, independent of which other collections are inserted. Do you feel always checking against the first element is better? That could open it up to test failures if we later decide to alter the ordering for some reason that I cannot yet foresee.

webknjaz replied:
I think either way is fine. But in general, I prefer tests to be as precise as possible to prevent some buggy behavior from squeezing in just because it has similar properties.

webknjaz commented on test/units/playbook/test_collectionsearch.py:
```diff
@@ -0,0 +1,38 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from ansible.playbook.collectionsearch import CollectionSearch
+
+import pytest
+
+
+def test_collection_static_warning(capsys):
+    """Test that collection name is not templated.
+
+    Also, make sure that users see the warning message for the referenced name.
+    """
+
+    collection_name = 'foo.{{bar}}'
+    cs = CollectionSearch()
+    assert collection_name in cs._load_collections(None, [collection_name])
+
+    std_out, std_err = capsys.readouterr()
+    assert '[WARNING]: "collections" is not templatable, but we found: %s' % collection_name in std_err
+    assert '' == std_out
```
@Shrews hope this works:
```suggestion
import logging

from ansible.playbook.collectionsearch import CollectionSearch

import pytest


def test_collection_static_warning(caplog):
    """Test that collection name is not templated.

    Also, make sure that users see the warning message for the referenced name.
    """

    collection_name = 'foo.{{bar}}'
    cs = CollectionSearch()
    with caplog.at_level(logging.WARNING, logger='ansible'):
        assert collection_name in cs._load_collections(None, [collection_name])

    expected_log_msg = '[WARNING]: "collections" is not templatable, but we found: %s' % collection_name
    assert caplog.record_tuples == [('ansible', logging.WARNING, expected_log_msg)]
```

Shrews replied:
I cannot make this work. caplog is always empty.

webknjaz replied:
Weird, maybe logging is disabled in tests. Oh, it looks like it's only activated if the config points to a writable path: https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/display.py#L79.

webknjaz replied:
But then it seems to default to `~` which should always be writable: https://github.com/ansible/ansible/blob/f921b1e3bd38d66c193b814c060ca1c6e97adc48/lib/ansible/config/base.yml#L810

webknjaz replied:
I guess it may be configurable by an env var but I don't know if `ansible-test` provisions it. Best to ask @mattclay.

Shrews replied:
Per our discussion, keeping use of capsys for now.

webknjaz commented on lib/ansible/playbook/collectionsearch.py:
```diff
@@ -41,4 +46,14 @@ def _load_collections(self, attr, ds):
+            if is_template(collection_name, env):
+                display.warning('"collections" is not templatable, but we found: %s, '
+                                'it will not be templated and will be used "as is".' % (collection_name))
+
```
This makes the linter explode:
```suggestion
```
_(https://app.shippable.com/github/ansible/ansible/runs/163982/1/tests)_

Shrews replied:
Not sure how that made it in, but fixed in my local repo. Will push when we sort out the test.

Shrews replied:
Fixed

Shrews committed:
Fix pep8 error

Shrews commented:
/rebuild_failed

sivel commented:
@Shrews can you add a changelog to `changelogs/fragments`?

Shrews commented:
> @Shrews can you add a changelog to `changelogs/fragments`?

Oops, sure thing.

Shrews committed:
Add changelog fragment

AlanCoding commented:
I could have sworn that templating collection names used to work, what were the specific circumstances that they were broken? Passing "namespace.name" via `-e` on the CLI certainly _seemed_ functional in 2.9.

s-hertel commented:
@AlanCoding Do you have a reproducer for it working on 2.9? I just tried on stable-2.9 and couldn't:
```
---
- hosts: localhost
  gather_facts: no
  collections:
    - "{{ collection }}"
  tasks:
    - module_name:
```
called with `-e collection='namespace.name'`

`ERROR! couldn't resolve module/action 'module_name'. This often indicates a misspelling, missing collection, or incorrect module path.`
